### PR TITLE
[utilities,plugins] speedup journal collection (v2)

### PR DESCRIPTION
**Pull request by champtar**
Instead of generating all the logs and tailing the last 100M, we get the first 100M of 'journalctl --reverse' that we then reverse again using our own implementation of tac.

To handle multiline logs we would need to use "tac -brs '^[^ ]'" that takes ~30s on 100M of logs when plain 'tac' takes ~0.3s. Our simple implementation in python takes 0.7s, and avoid an extra dependency.

On journalctl timeout we now get the most recents logs.

During collection logs are now buffered on disk, so we use 2xsizelimit. While running our tac we could actually truncate the source file to limit disk usage. Previously buffering was in RAM (also 2xsizelimit).

On my test server, logs plugin runtime goes from 34s to 9.5s.

Fixes https://github.com/sosreport/sos/issues/3615

---
Please place an 'X' inside each '[]' to confirm you adhere to our [Contributor Guidelines](https://github.com/sosreport/sos/wiki/Contribution-Guidelines)

- [X] Is the commit message split over multiple lines and hard-wrapped at 72 characters?
- [X] Is the subject and message clear and concise?
- [X] Does the subject start with **[plugin_name]** if submitting a plugin patch or a **[section_name]** if part of the core sosreport code?
- [X] Does the commit contain a **Signed-off-by: First Lastname <email@example.com>**?
- [X] Are any related Issues or existing PRs [properly referenced](https://docs.github.com/en/issues/tracking-your-work-with-issues/creating-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) via a Closes (Issue) or Resolved (PR) line?
- [X] Are all passwords or private data gathered by this PR [obfuscated](https://github.com/sosreport/sos/wiki/How-to-Write-a-Plugin#how-to-prevent-collecting-passwords)?

### CODE REVIEW

**Comment by pmoravec** [comment]
I can confirm VERY evident speedup on a tested RHEL8 system with:

```
# journalctl --disk-usage 
Archived and active journals take up 560.0M in the file system.
#
```

where current upstream executes `python3 bin/sos report -o logs --batch --build` in 18s and this PR shrinks it to 3s.

Some test is failing - can you @champtar review it or do you need some help? I might get to the failure later today  /tomorrow.

File: sos/report/plugins/__init__.py [hunk 1]
```diff
@@ -3083,10 +3084,18 @@ def add_journal(self, units=None, boot=None, since=None, until=None,
         if output:
             journal_cmd += output_opt % output
 
+        fname = journal_cmd
+        tac = False
+        if log_size > 0 and is_executable("head"):
+            journal_cmd = f"sh -c '{journal_cmd} --reverse | " \
+                "head -c {log_size*1024*1024}'"
```
Comment (pmoravec) [inline sos/report/plugins/__init__.py:3091]: Missing leading `f` as you use f-string :)

**Comment by champtar** [comment]
I'm off until the 27th so no rush for the review, and I was thinking about even implementing 'head' in Python

**Comment by pmoravec** [comment]
```
tests/report_tests/plugin_tests/defaults.py:40:            if cmd['exec'] == 'journalctl --no-pager  --unit cups':
```

needs to be updated to reflect changed command exec `"sh -c 'journalctl --no-pager  --unit cups --reverse | head -c 104857600'"` - I vote for `'journalctl --no-pager  --unit cups' in cmd['exec']` test.


The `2-tests/report_tests/plugin_tests/logs.py:JournalSizeLimitTest.test_journal_tailed_and_linked` failure: The PR breaks part of `tailit` functionality: when a command output is bigger than `--log-size` (or `--journal-size` here), sosreport looks like:

```
# ll /var/tmp/sosreport-pmoravec-rhel9-8675309-2024-12-18-dusvggk/sos_commands/logs/
total 24
-rw-r--r--. 1 root root    63 Dec 18 15:50 journalctl_--disk-usage
lrwxrwxrwx. 1 root root    51 Dec 18 15:50 journalctl_--no-pager -> ../../sos_strings/logs/journalctl_--no-pager.tailed
lrwxrwxrwx. 1 root root    58 Dec 18 15:50 journalctl_--no-pager_--boot -> ../../sos_strings/logs/journalctl_--no-pager_--boot.tailed
-rw-r--r--. 1 root root    82 Dec 18 15:50 journalctl_--no-pager_--boot_-1
-rw-r--r--. 1 root root 13355 Dec 18 15:50 ls_-alRh_.var.log
# ll /var/tmp/sosreport-pmoravec-rhel9-8675309-2024-12-18-dusvggk/sos_strings/logs/
total 24824
-rw-r--r--. 1 root root 10484923 Dec 18 15:50 journalctl_--no-pager_--boot.tailed
-rw-r--r--. 1 root root 10484923 Dec 18 15:50 journalctl_--no-pager.tailed
-rw-r--r--. 1 root root  4445393 Dec 18 15:50 var.log.messages-20241201.tailed
#
```

And `manifest.json` tracks that via `"truncated": true,`.

With this PR, I see:

```
# ll /var/tmp/sosreport-pmoravec-rhel9-8675309-2024-12-18-zagarel/sos_commands/logs/
total 20464
-rw-r--r--. 1 root root       63 Dec 18 15:50 journalctl_--disk-usage
-rw-r--r--. 1 root root 10460043 Dec 18 15:50 journalctl_--no-pager
-rw-r--r--. 1 root root 10460043 Dec 18 15:50 journalctl_--no-pager_--boot
-rw-r--r--. 1 root root       82 Dec 18 15:50 journalctl_--no-pager_--boot_-1
-rw-r--r--. 1 root root    22730 Dec 18 15:50 ls_-alZR_.var.log
#
```

That suggests the `journalctl` outputs are not truncated, which is misleading.

**Comment by pmoravec** [comment]
Hi @arif-ali and @TurboTurtle , is this approach worth to try? My answer is "carefully yes".

Hi @champtar , do you need some help with updating the tests?

**Comment by champtar** [comment]
@pmoravec cirrus-ci is broken right now, but this should be ready for review

File: sos/utilities.py [hunk 2]
```diff
@@ -275,8 +277,15 @@ def _check_poller(proc):
...
-        _output = open(to_file, 'w', encoding='utf-8')
+        if sizelimit:
+            # going to use HeadReader
+            _output = PIPE
+        elif tac:
+            # no limit but we need an intermediate file
+            _output = tempfile.TemporaryFile(dir=os.path.dirname(to_file))
+        else:
+            # pylint: disable=consider-using-with
+            _output = open(to_file, 'wb')
```
Comment (pmoravec) [inline sos/utilities.py:288]: (now nitpick) missing original `encoding='utf-8'`?

File: sos/utilities.py [hunk 3]
```diff
@@ -301,19 +320,25 @@ def _check_poller(proc):
...
                             'truncated': reader.is_full}
-            if to_file:
-                _output.close()
 
             # wait for Popen to set the returncode
             while p.poll() is None:
                 pass
 
+            if to_file:
+                if tac:
```
Comment (pmoravec) [inline sos/utilities.py:337]: Also nitpick now: why not merging into single `if to_file and tac:`?

**Comment by pmoravec** [comment]
Thanks for the new push. CI jobs restarted, seems good. I review the code now, will test functionality later (today or tomorrow, I assume).

File: sos/utilities.py [hunk 4]
```diff
@@ -496,8 +562,45 @@ def running(self):
...
+            buf = self.f_src.read(min(self.remaining, self.COPY_BUFSIZE))
+            if not buf:
+                break
+            self.f_dst.write(buf)
+            self.remaining -= len(buf)
+        self.f_src.close()
+        self.running = False
+
+    def get_contents(self):
+        return '' if not self.binary else b''
```
Comment (pmoravec) [inline sos/utilities.py:595]: note rather to myself: we return empty string here, despite the method is called when we return status 124, 126 or 127..? (I need to review behaviour in these use cases).

File: sos/utilities.py (hunk 2 above)
Comment (champtar) [reply sos/utilities.py:288]: Using binary mode (`wb`), encoding doesn't make sense

File: sos/utilities.py (hunk 3 above)
Comment (champtar) [reply sos/utilities.py:337]: There was something else in the `if to_file:` at some point (will fix on the next push)

**Comment by champtar** [comment]
Not in the PR, but I was thinking about making TailReader work with to_file at some point, as I think we don't use the 'output' that often

**Comment by pmoravec** [comment]
So far the PR survived various my corner-case tests I invented. I shall review the code tomorrow..

File: sos/utilities.py (hunk 2 above)
Comment (pmoravec) [reply sos/utilities.py:288]: Can't the binary mode negatively affect writing usual (textual) output of other commands..? It does not seem so, even encoding is preserved. But.. writing text command output using binary write is weird. While I understand with `tac=True` it is a must.

File: sos/utilities.py [hunk 5]
```diff
@@ -285,10 +294,20 @@
...
-                reader = AsyncReader(p.stdout, sizelimit, binary)
+            if to_file:
+                if sizelimit:
+                    if tac:
+                        _output = tempfile.TemporaryFile(
+                            dir=os.path.dirname(to_file)
+                        )
+                    else:
+                        # pylint: disable=consider-using-with
+                        _output = open(to_file, 'wb')
```
[the remark this answers is not in the document]
Comment (pmoravec) [reply sos/utilities.py:305]: Ditto - why binary mode for `tac=False`?

File: sos/utilities.py [hunk 6]
```diff
@@ -325,11 +350,52 @@ def _check_poller(proc):
...
+
+def tac_logs(f_src, f_dst, drop_last_log=False):
+    """Python implementation of the tac utility with support
+    for multiline logs (starting with space). It is intended
+    to reverse the output of 'journalctl --reverse'.
+    """
+    NEWLINE_B = b'\n'
+    NEWLINE_I = 10
+    SPACE_I = 32
+    f_src.flush()
```
Comment (pmoravec) [inline sos/utilities.py:369]: Nice one!

**Comment by pmoravec** [comment]
All up to one nitpick is good. I confirm great performance gain (and no visible performance penalty for journals having 100MB where the new approach additionally has to `tac` the data). Several corner cases checked, no issue found. Code seems good.

I like the idea, but would welcome another profound review of others as well - since we touch a frequently used code.

File: sos/utilities.py (hunk 3 above)
Comment (champtar) [reply sos/utilities.py]: Fixed

File: sos/utilities.py (hunk 6 above)
Comment (champtar) [reply sos/utilities.py:370]: Added a comment to make it explicit why it's mandatory to flush here

File: sos/utilities.py (hunk 2 above)
Comment (champtar) [reply sos/utilities.py:288]: I would reverse the question, why was it ever text mode ?
If the command returns binary, you don't want to use text mode.
I'm not even sure if it matters, as the command outputs directly to the file, I'm not sure text or binary has an impact here

Comment (pmoravec) [reply sos/utilities.py:288]: Outputs of commands collected by `sos` are usually of text content. So it makes sense to treat it as text and not binary.

Anyway I think the discussion here is rather theoretical - we are writing whole content directly to a file and such approach should provide same behaviour for text and binary formats (as far as I checked the text/bin differences).

**Comment by pmoravec** [comment]
I rerun the failed test - it should succeed, I rather smell some infra issue (and logs dont show any reason).

File: sos/utilities.py (hunk 2 above)
Comment (champtar) [reply sos/utilities.py:288]: Just tested, even for binary content it doesn't change anything:
```
import subprocess
f = open('test', 'wt')
subprocess.run(['printf', '\\0'], stdout=f)
subprocess.run(['printf', '\r\n'], stdout=f)
subprocess.run(['printf', '\r'], stdout=f)
```

```
$ hexdump -C test
00000000  00 0d 0a 0d                                       |....|
00000004
```

Looking with strace, we just `openat`
```
openat(AT_FDCWD</home/champtar>, "test", O_WRONLY|O_CREAT|O_TRUNC|O_CLOEXEC, 0666) = 7</home/champtar/test>
```

And then `dup2`
```
dup2(7</home/champtar/test>, 1</dev/pts/0<char 136:0>>) = 1</home/champtar/test>
```
The text vs binary only has an effect if we use the file with python function (read/write)

> Outputs of commands collected by sos are usually of text content. So it makes sense to treat it as text and not binary.

This function do support binary, so it's safer to handle everything as binary by default, as text is just a subset of binary (even if for now here it doesn't make any difference)

**Comment by champtar** [comment]
Thanks for the review @pmoravec !

**Review by pmoravec — APPROVED** [review]
Maybe it makes sense to squash some commits, but I am ok also with this granularity.

Looking forward for either @arif-ali or @TurboTurtle review.

**Comment by champtar** [comment]
@arif-ali or @TurboTurtle friendly ping

**Review by TurboTurtle — APPROVED** [review]
